File: go/42_TrappingRainWater.go
package main

func trap(height []int) int {
    if len(height) == 0 {
        return 0
    }
    var left, right, left_top, right_top, ret int
    right = len(height) - 1
    for left < right {
        if height[left] < height[right] {
            if height[left] > left_top {
                left_top = height[left]
            }
            ret += left_top - height[left]
            left++
        } else {
            if height[right] > right_top {
                right_top = height[right]
            }
            ret += right_top - height[right]
            right--
        }
    }
    return ret
}

File: python/42_TrappingRainWater.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        left = left_top = right_top = ret = 0
        right = len(height) - 1
        while left < right:
            if height[left] < height[right]:
                left_top = max(left_top, height[left])
                ret += left_top - height[left]
                left += 1
            else:
                right_top = max(right_top, height[right])
                ret += right_top - height[right]
                right -= 1
        return ret
